example app spec #65, #66

lucksus opened the pull request:
closes https://github.com/holochain/org/issues/65, https://github.com/holochain/org/issues/66, #2, #4  
# First iteration and starting point for further discussion and changes to the spec

In order to not engage in open-ended discussions, I've suggested to keep the scope of this first PR limited and then spec out more details and discuss changes in future PRs.

The scope of this first one should be: **getting a first testing round-trip with an actual app**.

What do we need for that?
- [x] Make Holochain read the changed DNA structure: https://github.com/holochain/holochain-rust/pull/253
- [x] Make hcdev create a bundle.json from this repository that can be processed without further manual changes (ongoing: https://github.com/holochain/holochain-cmd/pull/15/files)
- [x] Make zome functions actually callable - i.e. solve the `_dispatch` problem
  - [x] add a macro zome_fns! to the HDK that adds the wrapper code for de-/serializing complex parameters around the app developers function: https://github.com/holochain/hdk-rust/pull/6
  - [x] get rid of the `_dispatch` suffix since the macro can add that code directly in the same function and *dispatch* is the wrong word for what this does anyway: https://github.com/holochain/holochain-rust/pull/268
- [x] Move error code JSON to HDK: https://github.com/holochain/hdk-rust/pull/5
- [x] add comments to the parts that are only meant as stubs to invite further discussion (validation/genesis)
- [x] add JS/tape based test #2 (depends on https://github.com/holochain/holosqape/pull/5)




-----
*(old comment)*

I thought I'd start writing a simple app the way I would like it to be written, esp. with the assumption the DNA packager would be available already. So I've split the app up into several different files in a directory structure resembling the DNA tree. 

Several assumptions I applied:

* I assumed that the packager would take any ```manifest.json``` in each directory and add its contents to the object at the according place in the tree
* Only one WASM object per entry_type (other than what we assumed when writing the DNA spec) and the ribosome would infer the function names from the entry type name and the names (tags) of its links. (compare zomes/blog/entry_types/post/validation.rs with zomes/blog/entry_types/post/manifest.json)
* I've encountered a limitation of a design decision around links we've made earlier and had to change something / introduce a new concept: since we've replaced link entry types with a more succinct definition of links going off of an entry type, how would we create links from a system anchor like App.Key.Hash or App.DNA.Hash to an entry? Instead of only defining links going off of an entry type, I've added ```linked_from``` to zomes/blog/entry_types/post/manifest.json which is similar to ```links_to``` in that it implicitly creates a link type but different in that those links are based on a different entry and pointing to entries of the type in which the definition occurs. In this example links are being put onto ```App.Key.Hash``` and point to post entries. The validation code for those links would go into ``` fn validate_post_authored_posts() -> bool```. 

I also did:

* create a [repository for the HDK](https://github.com/holochain/hdk-rust) that is used here and add stubs so that I have a chance to compile this rust code

* build those files into a complete bundle.json file with hcdev and manually made last tweaks to get the desired output in dist/bundle.json

I would continue with (as @zippy mentioned in other tickets):

* adding JS/tape based tests in a ```tests``` directory that could be run in holoconsole
* setting up the CI so that it builds holoconsole and runs those tests

but would like to get feedback on this first since this feels like jumping ahead a bit and touches heavily on @sphinxc0re's work on the DNA packager and @thedavidmeister's work on the validate() ticket.

lucksus commented on zomes/blog/capabilities/main/main.rs:
```diff
@@ -0,0 +1,32 @@
+extern crate serde_json;
+use std::time::SystemTime;
+
+fn create_post(input: serde_json::Value) -> serde_json::Value {
```
I started off with both input and output being ```String```s and having a serde deserialize command as the first line in each function, but since we have to wrap these functions in a ```_dispatch()``` through our HDK anyway, why not use ```serde_json::Value``` here?

zippy replied:
Talking with @lucksus it becomes clear that the deeper issue here is that this also is about cross-language parameter handling (i.e. that works for any language calling in and out of the wasm), i.e. that inputs and outputs are always JSON. This makes sense to me because it means things are handled natively for typscript, and every other wams language will certainly have JSON handling.  Change request is that we document this in an ADR.

zippy replied:
The json!() thing is nice...

lucksus commented on zomes/blog/capabilities/main/main.rs:
```diff
@@ -0,0 +1,32 @@
+            "content": input["content"],
+            "date_created": SystemTime::now()
+        }
+    ));
```
This would then also be:
```rust
fn commit(type: String, entry: serde_json::Value) -> String
```

lucksus commented on zomes/blog/capabilities/main/manifest.json:
```diff
@@ -0,0 +1,42 @@
+    "membrane": "agent"
+  },
+
+  "fn_declarations": [
```
I actually want to change this to: ```abi```

zippy replied:
I get why ABI (and also what they use in Ethereum).  Hmmm.  But note that we also have signals that need to be delcared.

lucksus commented on zomes/blog/capabilities/main/manifest.json:
```diff
@@ -0,0 +1,42 @@
+          "outputs": [
+            {"name": "hash", "type": "string"}
+          ]
+        }
```
And as @sphinxc0re mentioned earlier, we should get rid of ```signature``` and have ```name```, ```inputs``` and ```outputs``` at the same level.
(I didn't want to introduce too many changes at once)

lucksus commented on zomes/blog/entry_types/post/manifest.json:
```diff
@@ -0,0 +1,16 @@
+    {
+      "base": "App.Key.Hash",
+      "tag": "authored_posts"
+    }
```
This implicitly defines link types that are being put onto ```App.Key.Hash``` pointing to posts under the tag "authored_posts". Can be retrieved like with ```get_links(agent, "authored_posts")```

zippy replied:
@lucksus and I talked about actually leaving anchors for the true anchor functionality that we want to merge into core, and convert this to `linked_from`

lucksus commented on zomes/blog/entry_types/post/validation.rs:
```diff
@@ -0,0 +1,23 @@
+struct Post {
+    content: String,
+    date_created: String
+}
```
This struct should be defined outside this file and then included here an in other WASM modules.

zippy commented on manifest.json:
```diff
@@ -0,0 +1,10 @@
+{
```
note that in @sphinxc0re 's implementation of holochain-cmd uses specific names for each file, not a generic name (manifest.json), so this needs to be reconciled on one side or the other.

zippy replied:
@lucksus and @sphinxc0re and @zippy spent some time talking about this, and the role of the packaging tool.  @lucksus and @zippy came down on the side of making the packager very simple and not really having a deep pre-understanding of the exact format of the directory and it's json contents, but rather simply to be able to take any tree and assemble what ever json files it finds there into a single json blob, with the smarts of also being able to compile the wasm code in finds there and add it into a "code" blob.  That way we can, in this initial dev phase use change the tree structure easily without having to recompile the packager tool.

zippy commented:
Also, as per the "I would continue with" items.  Lets add them as separate tickets.  I'm ok with you continuing on them for this sprint (especially the HDK and test one) but probably we can let the CI one wait.

lucksus commented:
Ok, this is not a WIP anymore. @zippy, I've implemented your change requests. I would like to merge this PR before sprint review and continue in separate tickets with the rest.

@sphinxc0re, I've built bundle.json using your `hcdev` but with several manual steps before and after. I've described that here in the readme. Not sure if I'm using it wrong and it could do all of that already. If not, it would be nice to get it to the point that all of that could be done automatically.

lucksus commented on README.md:
```diff
@@ -2,3 +2,31 @@
+  * remove hidden files and directories like `.DS_STORE` and `.git`
+  * remove `__META__` entries
+  * remove `README.me`
+  * change `zomes` and `capabilities` from an object to an array and move the name of the subitems into the subitems
```
I agree but left this for now and will change that in another PR because that is what is implemented now and I don't want to introduce to many changes at once.

lucksus commented:
@sphinxc0re, can you check if the current state of this PR is good for you? I've added .build files and removed the bullet points from the readme that talk about building wasm manually.

ddd-mtl commented:
The example is missing ErrorCode checks from calls to the hdk::* functions.
I don't understand why `genesis` is a required callback. Putting in an empty `genesis` for the sake of it seems pretty wasteful.
I understand the requirement for the `linked_from`, but I am worried that this name would confuse app devs into thinking that they should specify a `linked_from` for every `linked_to`. I don't have any alternative suggestions for the moment though.

ddd-mtl commented on zomes/blog/capabilities/main/code/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+
+[dependencies]
+serde_json = "1.0"
+hdk = { git = "https://github.com/holochain/hdk-rust"}
```
missing 'branch' attribute

Connoropolous replied:
Just like I mentioned in my other PR, git refs assume master branch, so this actually works (since hdk-rust is on master default branch).

ddd-mtl commented on zomes/blog/capabilities/main/code/src/lib.rs:
```diff
@@ -0,0 +1,47 @@
+    match hdk::get_entry(input["post_hash"].to_string()) {
+        Ok(entry) => json!({"post":  entry}),
+        Err(hdk::ErrorCode::HashNotFound) => json!({"post": {}, "error": "hash not found"}),
+        Err(hdk::ErrorCode::FunctionNotImplemented) => json!({"post": {}, "error": "get_entry not implemented yet!"}),
```
This should be done in the hdk (Have hdk::ErrorCode::to_json() or something like that)

thedavidmeister replied:
i agree that the json handling feels pretty manual for zome level code

ddd-mtl commented on zomes/blog/entry_types/post/validation/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+serde = "1.0"
+serde_json = "1.0"
+serde_derive = "1.0"
+hdk = { git = "https://github.com/holochain/hdk-rust"}
```
'branch'

Connoropolous replied:
Just like I mentioned in my other PR, git refs assume master branch, so this actually works (since hdk-rust is on master default branch).

thedavidmeister commented:
@ddd-mtl re: the link to and from, i think we just have to document some example hApp architectures that explain when/how you'd want to use each

@lucksus the links and other design decisions is where i really do like the idea of user stories, so we don't forget about use cases over time

thedavidmeister commented on zomes/blog/entry_types/post/validation/src/lib.rs:
```diff
@@ -0,0 +1,25 @@
+
+#[no_mangle]
+pub extern "C" fn validate_post_comments() -> bool {
+    true
```
is `true` required? or just an example?

lucksus replied:
-> Result<bool, ValidationError>
?

thedavidmeister commented on zomes/blog/capabilities/main/manifest.json:
```diff
@@ -0,0 +1,36 @@
+    "membrane": "agent"
+  },
+
+  "functions": [
```
just looking at https://github.com/faaslang/faaslang#faaslang-resource-definition for inspiration

they are trying to do almost the same thing as us... why don't we follow the standard, or something close to?

- support `description` for the function, inputs and outputs
- change `inputs` to `params`
- change `outputs` to `returns`
- support `defaultValue`
- support `context`
- support all the `type`s they support

differences:

- we can't support `bg` because we don't have "running in the background"
- we can't support `charge` because that's at the holo level, not holochain
- we don't need `format`
  - languages are at the zome level for us (or are they?)
  - we don't offer `async` (or do we?)
- we have an extra `type` here called `hash`

ddd-mtl replied:
Okay for me.
I do wonder how much we would want to have our own "community standard" more in line with P2P and Ceptr thinking.

lucksus replied:
As we discussed in today's meeting and how I've put it in the PR description, I think this should be a next PR so that we can merge this one fairly soon...

thedavidmeister commented on manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "properties": {
+      "test": "test"
+  }
+}
```
are any of these values required?

what should happen if i provide no root properties?

lucksus replied:
Yes, they *should* all be required, except for the "test" property, but even `properties` should be there, but could be empty.

thedavidmeister commented on zomes/blog/capabilities/hc_lifecycle/manifest.json:
```diff
@@ -0,0 +1,13 @@
+
+  "functions": [
+    {
+        "name": "genesis",
```
this isn't required by hc currently, `NotImplemented` is a pass

why would i ever want to add an empty genesis, other than to demonstrate that it exists?

lucksus replied:
As discussed today, this is not polished yet and mainly exists as a crystallization point for further discussion.

Imho, if there is a capability that is named `hc_lifecycle`, it should implement lifecycle functions like genesis, because we think of predefined system capabilities like this as traits that a zome can implement.
